### PR TITLE
Fset 640: Expire a candidate's online exercises

### PR DESCRIPTION
**Pull request by Lansalo**

### CODE REVIEW

**Comment by Lansalo** [comment]
I still would like to do some more work around the test coverage but it's probably low priority so I'll focus on the email template first.

**Comment by henricook** [comment]
@Lansalo Looks like there's some conflicts here, could we also have a more descriptive PR title please? E.g. "FSET 640 - Making a button to summon unicorns" Vs. just "FSET 640"

File: app/repositories/application/OnlineTestRepository.scala [hunk 1]
```diff
@@ -46,6 +46,8 @@ trait OnlineTestRepository {
 
   def insertOrUpdatePhase1TestGroup(applicationId: String, phase1TestProfile: Phase1TestProfile): Future[Unit]
 
+  def nextApplicationPendingExpiry: Future[Option[ExpiringOnlineTest]]
```
Comment (mgilgar) [inline app/repositories/application/OnlineTestRepository.scala]: Why don't we call the method "nextExpiringApplication"?

File: app/repositories/application/OnlineTestRepository.scala [hunk 2]
```diff
@@ -46,6 +46,8 @@ trait OnlineTestRepository {
```
Comment (mgilgar) [inline app/repositories/application/OnlineTestRepository.scala]: This is a general comment, will be have a separate repository for the different online tests: video interview, etc?

If yes, maybe we should renamed all things related to phase1 to something different maybe OnlineExercise or Phase1OnlineTest?

Anyway I would recommend to split everything in 3.

File: app/services/onlinetesting/OnlineTestExpiryService.scala [hunk 3]
```diff
@@ -42,16 +43,13 @@ class OnlineTestExpiryServiceImpl(
   newHeaderCarrier: => HeaderCarrier
 )(implicit executor: ExecutionContext) extends OnlineTestExpiryService {
 
-  private final val ExpiredStatus = "ONLINE_TEST_EXPIRED"
   private implicit def headerCarrier = newHeaderCarrier
 
   override def processNextExpiredTest(): Future[Unit] = {
```
Comment (mgilgar) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: I think if the name of the method is "processNextExpiredTest", then this line:

otRepository.nextApplicationPendingExpiry

should be:
otRepository.nextApplicationExpiredTest

nextApplicationPendingExpiry is supposed to return an application (by the name), if it is an online tests what is returning, it should say: nextExpiredOnlineTest.

File: app/services/onlinetesting/OnlineTestExpiryService.scala [hunk 4]
```diff
@@ -42,16 +43,13 @@ class OnlineTestExpiryServiceImpl(
...
-    // TODO FAST STREAM FIX ME
-    Future.successful(Unit)
-    //otRepository.nextApplicationPendingExpiry.flatMap {
-    //  case Some(expiringTest) => processExpiredTest(expiringTest)
-    //  case None => Future.successful(())
-    //}
+    otRepository.nextApplicationPendingExpiry.flatMap {
+      case Some(expiringTest) => processExpiredTest(expiringTest)
```
Comment (mgilgar) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: if we get an expiringTest, as I wrote before, it is weird that the method is called nextApplication....

File: app/services/onlinetesting/OnlineTestExpiryService.scala [hunk 5]
```diff
@@ -67,7 +65,7 @@ class OnlineTestExpiryServiceImpl(
     }
 
   override def commitExpiredStatus(expiringTest: ExpiringOnlineTest): Future[Unit] =
-    applicationRepository.updateStatus(expiringTest.userId, ExpiredStatus).map { _ =>
```
Comment (mgilgar) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: In this case I think it should say commietExpiredProgressStatus

File: app/repositories/application/OnlineTestRepository.scala (hunk 2 above)
Comment (Lansalo) [inline app/repositories/application/OnlineTestRepository.scala]: Don't think will have time given the deadline. Do you want to track this somewhere?

File: app/repositories/application/OnlineTestRepository.scala (hunk 1 above)
Comment (Lansalo) [inline app/repositories/application/OnlineTestRepository.scala]: No particular reason apart from the fact that that was the name in FastTrack.
Happy to change it though.

File: app/services/onlinetesting/OnlineTestExpiryService.scala (hunk 3 above)
Comment (Lansalo) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: You mean "nextExpiringApplication" like in the comment above right?
I assume so as it's the same method and also at this stage the application is not really expired yet.

File: app/services/onlinetesting/OnlineTestExpiryService.scala (hunk 4 above)
Comment (Lansalo) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: Done

File: app/services/onlinetesting/OnlineTestExpiryService.scala (hunk 5 above)
Comment (Lansalo) [inline app/services/onlinetesting/OnlineTestExpiryService.scala]: Done

**Pull request merged**